sirepo/template/template_common.py

```diff
@@ -119,6 +120,170 @@ def __scale_model(self, name, model, is_native):
         return model
 
 
+class JupyterNotebook(object):
+    """Make a notebook
+    """
+
+    _CELL_TYPES = ('code', 'markdown')
+    _HEADER_CELL_INDEX = 0
+    _IMPORT_HEADER_CELL_INDEX = 1
+    _IMPORT_CELL_INDEX = 2
+    _PYPLOT_STYLE_MAP = PKDict(
+        line='-',
+        scatter='.',
+    )
+
+    @classmethod
+    def _base_dict(cls):
+        return PKDict(
+            cells=[],
+            nbformat=4,
+            nbformat_minor=4,
+            metadata=PKDict(
+                kernelspec=PKDict(
+                    display_name='Python 3',
+                    language='python',
+                    name='python3'
+                ),
+                language_info=PKDict(
+                    codemirror_mode=PKDict(
+                        name='ipython',
+                        version=3
+                    ),
+                    file_extension='.py',
+                    mimetype='text/x-python',
+                    name='python',
+                    nbconvert_exporter='python',
+                    pygments_lexer='ipython3',
+                    version='3.7.2'
+                )
+            )
+        )
+
+    @classmethod
+    def _dict_to_kwargs_str(cls, d):
+        d_str = ''
+        for k in d:
+            v = f'\'{d[k]}\'' if isinstance(d[k], pykern.pkconfig.STRING_TYPES) \
+                else f'{d[k]}'
+            d_str += f'{k}={v},'
+        return d_str
+
+    def __init__(self, sim_type, data):
+        self.data = data
+        self.notebook = JupyterNotebook._base_dict()
+        self.imports = PKDict()
+        self.widgets = []
+
+        # cell 0
+        self.add_markdown_cell(
+            [
+                '# {} - {}'.format(sim_type, data.models.simulation.name),
+            ]
+        )
+        # cell 1
+        self.add_markdown_cell(
+            ['## Imports',]
+        )
+        # cell 2
+        self.add_code_cell([])
+
+        super(object, self).__init__()
+
+    def add_cell(self, cell_type, source_strings, hide=False):
+        assert cell_type in self._CELL_TYPES, 'Invalid cell type {}'.format(cell_type)
+        cell = PKDict(
+            cell_type=cell_type,
+            metadata={},
+            source=[s + ('\n' if s[-1] != '\n' else '') for s in source_strings]
+        )
+        if hide:
+            cell.metadata['jupyter'] = {
+                'source_hidden': 'true'
+            }
+        self.notebook.cells.append(cell)
+
+    def add_code_cell(self, source_strings, hide=False):
+        self.add_cell('code', source_strings, hide=hide)
+
+    # {<pkg>: [sub_pkg]}
+    # just merge?
+    def add_imports(self, pkg_dict):
```

Comment:
I don't understand this complexity. _update is only called here afaict. Why not just append strings if not in the list and do it lterally. I think add_imports could look like this?

```py
def add_imports(self, module, package=''):
   
    s = (package and f'from {package} ') + f'import {module}\n'
    x = self.notebook.cells[self._IMPORT_CELL_INDEX].source
    if s not in x:
        x.append(s)
```

Reply:
0242d7d

Got rid of this entirely in favor of using code cells, but will revisit if that becomes ugly